### PR TITLE
Use detached lifetime for stats actor

### LINKED ISSUE #25237: [Core] Dangling actor handle held at Ray client
TL;DR user tried to use an actor at Ray client (via the actor handle it kept) but the actor was already gone in Ray cluster, resulting in KeyError.

**What happened**
The user was running a series of tune trials, and they restarted Ray cluster and reconnected for each trial from Ray client.
The first trial can work fine but in the second trial it crashed on KeyError, when it tried to load data via ray.data.read_parquet().

**Hypothesis**
When we create Dataset we use a stats actor to keep execution stats. We reuse the same stats actor for all Datasets created by the same process. This is done via keeping the actor handle in a global variable after the first time the actor was created.

The issue is the process (Ray client) doesn't reset or restart when it restarts the Ray cluster in each trial. So after the first trial and the Ray cluster restarted, the actor was already gone in the cluster. The actor handle kept in Ray client process is dangling.

**Verification**
First, we verified that the actor handle was indeed the same as the one used in the first trial.
Second, we verified that it will work, if we manually nulled out the stats actor handle and re-created it in each trial.

**Solutions**
- Option 1: keep a node_id together with the actor handle, and reuse stats actor only if Datasets are created under same node_id
- Option 2: use the client_id, which is uuid per connection, to detect the change and reuse the stats actor only if Datasets are created under same client_id

### PR DESCRIPTION
**Pull request by jianoaix**
## Why are these changes needed?
The actor handle held at Ray client will become dangling if the Ray cluster is shutdown, and in such case if the user tries to get the actor again it will result in crash. This happened in a real user and blocked them from making progress.

This change makes the stats actor detached, and instead of keeping a handle, we access it via its name. This way we can make sure re-create this actor if the cluster gets restarted.

## Related issue number

Closes #25237

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/impl/stats.py
Comment (ericl) [inline python/ray/data/impl/stats.py:118]: Hmm I don't remember why we did it this way, would it be better to use `actor.options(get_if_exists=True, name="_dataset_stats_actor").remote()` instead to get-or-create the actor?

Comment (jianoaix) [reply python/ray/data/impl/stats.py:118]: I thought about changing to this option (getting by name) as well, but it seems not meet the requirement. 
So the requirement here is that this actor is reused across datasets created by this same process. For the getting by name approach:
- It may work for the same dataset, but after it's completed and the actor refcount goes to zero, the datasets created later will not be able to get that same actor (have to recreate).
- Alternatively, we may create a detached actor, but in that case it will be shared even across different driver processes.

Keeping a reference here and clearing it upon new connection or new driver is essentially what this PR does to serve this requirement.

Comment (ericl) [reply python/ray/data/impl/stats.py:118]: Hm, sharing across drivers should be ok right? That should be uncommon case anyway.

Comment (jianoaix) [reply python/ray/data/impl/stats.py:118]: If that's not a concern, then getting by name is indeed simpler.

**Review by ericl — APPROVED** [review]

File: python/ray/data/impl/stats.py
Comment (ericl) [inline python/ray/data/impl/stats.py:116]: namespace="_dataset"

**Review by ericl — CHANGES_REQUESTED** [review]
Ah also need to set the namespace, otherwise you'll leak one actor per job.

**Comment by clarkzinzow** [comment]
@ericl @jianoaix So we can assume that the `get_if_exists=True` path is ~as cheap as the global actor handle cache since the core worker should cache actor handle fetches, right? It looks as if that path should only involve an RPC on the first fetch for a given worker process.

**Comment by ericl** [comment]
Hmm that's a good question, we should validate this for sure prior to merging.

**Comment by jianoaix** [comment]
Compared to the data loading of read task, one RPC seems a small cost?

Do we have a test to run the impact of this?

**Comment by ericl** [comment]
Simple enough to run a trivial dataset workload with small blocks before/after this PR. Maybe like 10000 blocks map batches?

**Comment by jianoaix** [comment]
Tried a simple test like this:
```
    total_time = 0
    for _ in range(16):
        start_time = time.time()
        ds = ray.data.range(100000, parallelism=10000)
        ds.map_batches(lambda x: x)
        total_time = time.time() - start_time
    print("mean time:", total_time / 16)
```
On a local cluster with 8 nodes and 1 cpu/node:
```
def build_cluster(num_nodes, num_cpus):
    cluster = Cluster()
    for _ in range(num_nodes):
        cluster.add_node(num_cpus=num_cpus)
    cluster.wait_for_nodes()
    return cluster

cluster = build_cluster(8, 1)
```

With this PR: mean time: 2.742537647485733
Without this PR: mean time: 2.706667184829712
The difference is 1.33%, which is small given the block is very small (just 10 ints). But the nodes are all on laptop so the RPC might be cheaper than real cluster.

**Review by ericl — APPROVED** [review]
Good enough for me!

File: python/ray/data/impl/stats.py
Comment (clarkzinzow) [reply python/ray/data/impl/stats.py:116]: So this stats actor will never be cleaned up for the lifetime of the Ray cluster, and the read stats for each `Dataset` will never get cleaned up within the actor, which is a bit of a leaky lifecycle. This seems fine for now to unblock the Ray Client use case, but we should probably open a P2 to improve the stats actor lifecycle, or eliminate the stats actor all together, if possible.

**Review by clarkzinzow — APPROVED** [review]

File: python/ray/data/impl/stats.py
Comment (jianoaix) [reply python/ray/data/impl/stats.py:116]: My previous code was made quite close to the existing way. I think the feedback here is it's a bit complicated AND detached actor isn't a concern to use. It may make sense to have a new kind of lifetime between refcounted and detached, e.g. per-job lifetime actor.

**Comment by jianoaix** [comment]
Microbenchmark:
```
    start_time = time.time()
    for _ in range(1000):
        ah = ray.data.impl.stats._get_or_create_stats_actor()
    print("mean time to get:", (time.time() - start_time) / 1000)
```

Before: 1.4783143997192383e-05 (sec)
After: 0.0005355322360992432 (sec)
Diff: 36x increase

**Comment by ericl** [comment]
Some broken tests. Are there any blocking concerns to revisit the design? The microbench result seems expected / fine.

**Comment by jianoaix** [comment]
This was to fix an issue from a main user. It was a blocker for them but then they were satisfied with a hack, so this got deprioritized.
This user is now planning to upgrade to 1.13, so this will be needed to remove that hack. I'll revive this PR.

**Comment by jianoaix** [comment]
Slightly updated to make the call to get actor once per Dataset, rather than once per task.

**Comment by jianoaix** [comment]
I was looking at the test failures, but it turned out they were already on the flaky test list. So the PR is ready to review/merge.

**Comment by jianoaix** [comment]
@ericl @clarkzinzow ptal, thanks

**Review by clarkzinzow — APPROVED** [review]
LGTM, although I am a bit concerned about the leaky lifecycle of a long-lived detached actor. This stats actor will live for the lifetime of the cluster, and read stats won't be cleaned up for the lifetime of the cluster.

This could be done as a follow-up (and shouldn't block merging this), but what do you think about adding best-effort clean-up of these read stats when `DatasetStats` is destructed?

```python
@ray.remote(num_cpus=0)
class _StatsActor:
    # ...

    def clear(self, stats_uuid: str):
        self.metadata.pop(stats_uuid, None)
        self.last_time.pop(stats_uuid, None)
        self.start_time.pop(stats_uuid, None)

class DatasetStats:
    def __del__(self):
        if self.needs_stats_actor:
            self.stats_actor.clear.remote(self.stats_uuid)
```

**Comment by jianoaix** [comment]
Not sure it'll work, DatasetStats as a Python/local object can have multiple instances in cluster, this cannot clean up the entry in actor for each destruction.

**Comment by clarkzinzow** [comment]
@jianoaix Ah good point, yeah I forgot that stats can be sent around to other tasks. It doesn't seem like there's a good garbage collection point at the moment. 🤔

File: python/ray/data/_internal/stats.py
Comment (clarng) [inline python/ray/data/_internal/stats.py:117]: nit

```suggestion
def _create_stats_actor():
```

File: python/ray/data/_internal/lazy_block_list.py
Comment (clarng) [inline python/ray/data/_internal/lazy_block_list.py:507]: do we need the null check since it is already using get_if_exists=True

Comment (jianoaix) [reply python/ray/data/_internal/lazy_block_list.py:507]: This is actually the cached actor handle in this class which was initialized to None, so it's None for the first call here.

**Comment by scv119** [comment]
Do we need cherry pick this to 2.0.0 release branch?

**Comment by jianoaix** [comment]
> Do we need cherry pick this to 2.0.0 release branch?

We haven't heard any issues other than the user who originally reported this (holding a ray client, and shutdown/restarting cluster for multiple trials), so we probably do not need to pick it.

**Comment by jianoaix** [comment]
Synced to head and CI passed, @clarkzinzow

**Comment by ericl** [comment]
What if we just have a FIFO queue of stats? Like the most recent 10000
Dataset stats, which should suffice for almost everyone, but safeguard
against any worst case OOMs like if you're running datasets in a while loop
or something.

**Pull request merged**
